README.md
# pjbserver-tools
[![PHP Version](http://img.shields.io/badge/php-5.5+-ff69b4.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)
[![PHP Version](https://img.shields.io/badge/php-7.1+-ff69b4.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)
[![PHP Version](https://img.shields.io/badge/php-8.0+-ff69b4.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)
[![Build Status](https://travis-ci.org/belgattitude/pjbserver-tools.svg?branch=master)](https://travis-ci.org/belgattitude/pjbserver-tools)
[![codecov](https://codecov.io/gh/belgattitude/pjbserver-tools/branch/master/graph/badge.svg)](https://codecov.io/gh/belgattitude/pjbserver-tools)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/belgattitude/pjbserver-tools/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/belgattitude/pjbserver-tools/?branch=master)
[![Latest Stable Version](https://poser.pugx.org/belgattitude/pjbserver-tools/v/stable.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)
[![Total Downloads](https://poser.pugx.org/belgattitude/pjbserver-tools/downloads.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)
[![License](https://poser.pugx.org/belgattitude/pjbserver-tools/license.svg)](https://packagist.org/packages/belgattitude/pjbserver-tools)

The `pjbserver-tools` package currently provides an easy installable php java bridge standalone server.
See the [soluble/japha](https://github.com/belgattitude/soluble-japha) and [php-java-bridge](https://github.com/belgattitude/php-java-bridge)
projects to get more info about php/java integration.

> **Note** that the `pjbserver-tools` standalone server utility
> was made for easy composer installation when developing or testing.
>
> **For production, building a customized PHPJavaBridge server is
> really easy. Check the recommended [php/java/bridge installation notes](http://docs.soluble.io/soluble-japha/install_server/)**


## Use cases

The java bridge standalone server can be used as an alternative to a regular bridge installation
for php/java integration while keeping things simple for development, unit testing or small projects.

## Features

- Easy setup of a PHP Java bridge standalone server (*nix).
- Console commands to control the server (start/stop/restart/status).
- API library to customize the behaviour.
- Includes compiled [JavaBridge.jar](./resources/pjb713_standalone/JavaBridge.jar) 7.1.3 file.

## Requirements

- PHP 5.5+, 7.0 or HHVM >= 3.2.
- Linux/Unix *(Mac OSX 10.1+ reported working)*.
- Java 1.7+ (see [install instructions](./doc/install_java.md)).

## Installation

> **WARNING** The phpjavabridge server is not supposed to be run on a public facing server
> and its use should be limited to interactions on the same host/network with the php client.
> Do not run it as root neither as it exposes the JVM methods through the network that
> could be remotely exploited. Do not start it as root neither.

Depending on your needs you can use the pjserver-tools in two ways.

1. Option 1: Composer install

   You can easily add the pjbserver-tools to your existing [composer](http://getcomposer.org/) project.

    ```console
    $ composer require belgattitude/pjbserver-tools
    ```

    It can also be added to your development dependencies (replace `require` by `require --dev` in the
    previous command). Very helpful if you intend to test with Travis...

    Test a command line.

    ```console
    ./vendor/bin/pjbserver-tools pjbserver:status ./vendor/belgattitude/pjbserver-tools/config/pjbserver.config.php.dist -vvv
    ```


2. Option 2: Console, clone the repo.

   First create a path on your filesystem that will hold the server install.

   ```console
   $ mkdir -p /my/path/pjbserver-tools
   $ cd /my/path/pjbserver-tools
   ```

   Clone the repository and use run [composer](http://getcomposer.org) update.

   ```console
   $ git clone https://github.com/belgattitude/pjbserver-tools.git
   $ cd pjbserver-tools
   $ composer update
   ```

   Test a command line

   ```
   ./bin/pjbserver-tools pjbserver:status ./config/pjbserver.config.php.dist -vvv
   ```


## Usage

### Command line

Command line depends on your install method (composer or clone/download).

- With composer the location of the binary is `./vendor/bin/pjbserver-tools` and the default
  config is located in `./vendor/belgattitude/pjbserver-tools/config/pjbserver.config.php.dist`.

- With the clone method, binary is `./bin/pjbserver-tools` and default config is `./config/pjbserver.config.php.dist`.

*For clarity, the documentation of console commands is based on the clone method. Simply
replace your path whenever needed.*

You can use the commands
`pjbserver:start`, `pjbserver:stop`, `pjbserver:restart`, `pjbserver:status` followed
by the `pjbserver.config.php` file to control or query the server status.


```console
$ ./bin/pjbserver-tools pjbserver:start -vvv ./config/pjbserver.config.php.dist
$ ./bin/pjbserver-tools pjbserver:stop -vvv ./config/pjbserver.config.php.dist
$ ./bin/pjbserver-tools pjbserver:restart -vvv ./config/pjbserver.config.php.dist
$ ./bin/pjbserver-tools pjbserver:status -vvv ./config/pjbserver.config.php.dist

$ # for listing the java cli command issued :
$ ./bin/pjbserver-tools pjbserver:reveal ./config/pjbserver.config.php.dist
```


If you use the [./config/pjbserver.config.php.dist](./config/pjbserver.config.php.dist) config file, the server will start on port ***8089***.

Feel free to create a local copy of this file and adapt for your usage :

```console
$ cp ./config/pjbserver.config.php.dist /my/path/pjbserver.config.php
```

*Note that the -v, -vv, -vvv option in the command line allows to define the verbosity level of the scripts.*

### Controlling via the API

Command line is good, but API gives a little more control especially good when setting unit tests and CI.

Here's a little example:

```php
<?php

use PjbServer\Tools\StandaloneServer;
use PjbServer\Tools\StandaloneServer\Config;

$tcp_port = 8089;

$config = new Config([
    // Port on which php java bridge server listen (required)
    'port' => $tcp_port,

    /**
     * Location of log and pid files...
     * Defaults is to put them in the project 'pjbserver-tools/var/...' directory
     * which is fine for unit testing, but to prevent loosing those files
     * set a safe directory (not /tmp as it might be cleared by the OS)
     */
    //'log_file'   => "/my/path/var/pjbserver-port${tcp_port}.log",
    //'pid_file'   => "/my/path/var/pjbserver-port${tcp_port}.pid",


    // Optional but often more than useful
    'classpaths'  => [
          '/my/path/*.jar',
          '/another/path/mylib.jar'
    ],

    // Standalone server tuning
    // Number of threads for standalone server is 50, increase if needed
    //'threads'    => 50,

    // Java binary
    // change location if you like, for example
    // /usr/lib/jvm/java-8-oracle/bin/java
    'java_bin' => 'java',

    /**
     * Location of the JavaBridge.jar,
     * Default is to use the default (included) one
     * available in pjbserver-tools/resources/pjb713_standalone/JavaBridge.jar
     */
    //'server_jar' => "/my/path/pjb713_standalone/JavaBridge.jar",
]);

$server = new StandaloneServer($config);

try {
    $server->start();
} catch(\Exception $e) {
    // Exception message
    echo $e->getMessage();
    // Server output logs
    echo $server->getOutput();
    die();
}

echo "Started: " . ($server->isStarted() ? 'yes' : 'no') . PHP_EOL;
echo "Running: " . ($server->isProcessRunning() ? 'yes' : 'no') . PHP_EOL;
echo "Pid    : " . $server->getPid() . PHP_EOL;

// Stopping the server

$server->stop();

```

You can also inject any PSR-3 compatible logger to the `StandaloneServer`.

```php
// any PSR-3 compatible logger
$logger = new \Psr\Log\NullLogger();
$server = new StandaloneServer($config, $logger);

```

## Configuration

The dist config file [./config/pjbserver.config.php.dist](https://github.com/belgattitude/pjbserver-tools/blob/master/config/pjbserver.config.php.dist)
contains the default parameters used in console mode.

### Parameters

| Key            | Type   | Comment                                          |
|----------------|--------|--------------------------------------------------|
| `port`         | int    | TCP port on which standalone server listen       |
| `classpaths`   | array  | Java additional classpaths                      |
| `threads`      | int    | Server max number of threads                     |
| `java_bin`     | string | Java binary executable (with or without path)    |
| `server_jar`   | string | Path to the JavaBridge.jar file                  |
| `log_file`     | string | Path to the standalone server log file           |
| `pid_file`     | string | Path to the standalone pid file                  |

Some considerations:

- When choosing a `port`, ensure it's not available publicly (security).
- The default config set `log_file` and `pid_file` in the ./var directory, please
  change the default location to your data directory.
- Avoid storing `log_file` and `pid_file` in the global temp directory '/tmp' as it might
  be cleared by the OS at anytime.

### Classpath configuration

Whenever you need to add some java libraries, simply edit the configuration file and look for the
`classpaths` option and add the required jar files.

As an example:

```php
<?php

return [
    'port'       => 8089,
    'classpaths' => [
        '/my/path/autoload/mysql-connector.jar',
        '/my/autoload_path/*.jar'
    ],
];
```

Don't forget to restart the standalone server to reflect the changes.

*Using wildcard* `/my/path/*.jar` *is possible but should be used with care. All matching files will be appended to classpath
by passing arguments in a shell exec. Limits exists...*

## Debugging

Some useful commands to watch, debug and eventually kill java standalone server process

### Getting the status (running/not running)

```console
$ ./bin/pjbserver-tools pjbserver:status -vvv ./config/pjbserver.config.php.dist
```

### Reveal the issued command

```console
$ ./bin/pjbserver-tools pjbserver:reveal -vvv ./config/pjbserver.config.php.dist
```

For example, the issued command the default config can be

```console
$ java -cp "/xxx/pjbserver-tools/resources/pjb713_standalone/JavaBridge.jar" -Dphp.java.bridge.daemon="false" -Dphp.java.bridge.threads=50 php.java.bridge.Standalone SERVLET:8089
```

### Process management

If for any reason the server cannot be stopped through the console, you can
lookup the process through the command line.

```console
$ # Searching by listening port
$ netstat -nlp | grep :<port>   # might require sudo if server started as root

$ # Searching by name
$ ps ax | grep JavaBridge.jar

$ # Searching by custom filter
$ pgrep -f "(.*)java(.*)JavaBridge.jar(.*)SERVLET:8089"
```

You can kill the process:

```
$ kill <pid_standalone_server>
```
## FAQ

### How to start automatically at boot

Check for "supervisord" on google, you'll find some recipes but the
preferred method is to deploy on Tomcat, see:

- [pjb-starter-springboot](https://github.com/belgattitude/pjb-starter-springboot/) to build you own production, self-container or deployable servlet or
- [php-java-bridge](https://github.com/belgattitude/php-java-bridge) project.

For alternatives to pjbserver-tools standalone.

## Coding standards

* [PSR 4 Autoloader](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader.md)
* [PSR 3 Logger interface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md)
* [PSR 2 Coding Style Guide](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md)
* [PSR 1 Coding Standards](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md)


CHANGELOG.md
# CHANGELOG

## 4.1.0 (2021-04-25)

- Fix: for some slow CI, ensure the port is ready when starting [#11](https://github.com/belgattitude/pjbserver-tools/pull/11)
- Add support for soluble-japha ^3.0 in [#11](https://github.com/belgattitude/pjbserver-tools/pull/11)

## 4.0.0 (2021-04-19)

- Support for php 8.0 and php 7.4
- Removed support for php 7.1, 7.2 and 7.3

## 3.1.0 (2019-12-21)

- Support for symfony/console ^5.0

## 3.0.0 (2019-02-18)

- No breaking changes. Just more type checks and support for symfony/console ^4.0.

### Added

- Support for symfony/console ^4.0

### Changed

- PHP 7.1 with strict_types.
- PHP 7.2 and 7.3 tested.
- PHPUnit upgraded to 7.5

### Removed

- Drop support for hhvm (travis).
- Drop support for soluble/japha < 2.5.0.
- Drop support for symfony/console < 3.0.0.

## 2.2.0 (2017-09-18)

### Updated

- Updated standalone server to [php-java-bridge 7.1.3](https://sourceforge.net/projects/php-java-bridge/files/Binary%20package/),
  you can still explicitly set the 6.2.1 version in the config `'server_jar' => "/my/path/pjb621_standalone/JavaBridge.jar"`

## 2.1.2 (2017-01-18)

### Improved

- Updated documentation

### Removed

- Removed obsolete resources in `./tools` directory

## 2.1.1 (2017-01-15)

### Changed

- Updated embedded `JavaBridge.jar` from [belgattitude/php-java-bridge](https://github.com/belgattitude/php-java-bridge) latest build.
- Preliminary support for pre/post composer hooks

### Fixed

- [#7](https://github.com/belgattitude/pjbserver-tools/pull/7) fix markdown in README.md
- HHVM badge updated for testing HHVM 3.9


## 2.1.0 (2016-10-27)

### Changed

- Refactored system process handling (`PjbServer\Tools\System\Process`)
- Console, start and stop commands messages :
  - pjbserver:start inform if the server was already started.
  - pjbserver:stop inform if the server was already stopped.

## 2.0.6 (2016-10-24)

### Added

- Nothing

### Changed

- Reworked default configuration, 'log_file', 'pid_file' and 'server_jar' are now
  commented and use defaults.

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing


## 2.0.5 (2016-10-24)

### Added

- Experimental support for Mac OSX
- Added PHP7.1 tests in travis

### Deprecated

- Substitution of {base_dir} and {tcp_port} variables in config variables does not prove
  a good option, they'll be removed for the v3.0.0 release. See
  [#4](https://github.com/belgattitude/pjbserver-tools/issues/4)

### Removed

- Nothing.

### Fixed

- [#3](https://github.com/belgattitude/pjbserver-tools/issues/3) fix an
  incompatibility with OSX using 'ps --no-headers' to check if pid is running.
  Replaced by 'kill -0' which should be more compatible.


## 2.0.4 (2016-09-06)

- Added possibility to modify standalone server number of threads from config file.
- New command to reveal internal cli command used to start the server
  see ./bin/pjbservertools pjbserver:reveal ./config/pjbserver.config.php.dist

## 2.0.3 (2016-07-04)

- Added composer commands aliases with default configuration file
  from ./config/pjbserver/config.php.dist and verbosity level to debug.
  Will start on port 8089 by default.
    - composer pjbserver:start
    - composer pjbserver:stop
    - composer pjbserver:restart
    - composer pjbserver:status

- Added console command to get the status of the server
  - ./bin/pjbservertools pjbserver:status <config_file>
- Added console logger to all commands:
  - To get more verbose message try the usual command with -v, -vv or -vvv
    i.e: ./bin/pjbservertools -vvv pjbserver:start ./config/pjbserver.config.php.dist

## 2.0.2 (2016-07-03)

- Fix travis builds (no notable changes)

## 2.0.1 (2016-07-03)

- Fix for multiple jar entries (*.jar)

## 2.0.0 (2016-07-03)

- Refactored usage, config params must now be passed in the `StandaloneServer\Config($params)` object.
- Support for PSR-3 logger interface.
- Console commands added (symfony/console)
  - ./bin/pjbservertools pjbserver:start <config_file>
  - ./bin/pjbservertools pjbserver:stop <config_file>
  - ./bin/pjbservertools pjbserver:restart <config_file>



## 1.1.0 (2016-07-02)

- PHP 5.5 as minimum required version.
- Improved unit testing and documentation.
- Added method `StandaloneServer::isProcessRunning()`.



src/PjbServer/Tools/StandaloneServer/Config.php
<?php

declare(strict_types=1);

namespace PjbServer\Tools\StandaloneServer;

use PjbServer\Tools\Exception;

/*
  java -Djava.awt.headless="true"
  -Dphp.java.bridge.threads=50
  -Dphp.java.bridge.base=/usr/lib/php/modules
  -Dphp.java.bridge.php_exec=/usr/local/bin/php-cgi
  -Dphp.java.bridge.default_log_file=
  -Dphp.java.bridge.default_log_level=5
  -Dphp.java.bridge.daemon="false"
  -jar JavaBridge.jar
 * sudo netstat -anltp|grep :8089
 */

class Config
{
    /**
     * Default configuration options.
     *
     * @var array<string, mixed>
     */
    protected $default_config = [
        'java_bin' => 'java',
        'server_jar' => '{base_dir}/resources/pjb713_standalone/JavaBridge.jar',
        'log_file' => '{base_dir}/var/pjbserver-port{tcp_port}.log',
        'pid_file' => '{base_dir}/var/pjbserver-port{tcp_port}.pid',
        'classpaths' => [],
        'threads' => 50
    ];

    /**
     * Internal configuration array.
     *
     * @var array<string, mixed>
     */
    protected $config;

    /**
     * Constructor.
     *
     * <code>
     *
     * $params = [
     *      // Port (required)
     *      'port' => 8089,
     *
     *      // Classpath autoloads (optional)
     *      'classpaths' => [
     *          '/my/path/to_specific/jar_file.jar',
     *          '/my/path/to_all_jars/*.jar'
     *      ],
     *
     *      'threads' => 50,
     *
     *      // Defaults (optional)
     *      'java_bin'   => 'java',
     *      'server_jar' => '{base_dir}/resources/pjb621_standalone/JavaBridge.jar',
     *      'log_file'   => '{base_dir}/var/pjbserver-port{tcp_port}.log',
     *      'pid_file'   => '{base_dir}/var/pjbserver-port{tcp_port}.pid'
     *
     * ];
     * $config = new StandaloneServer\Config($params);
     * </code>
     *
     * @throws Exception\InvalidArgumentException
     *
     * @param array<string, mixed> $config
     */
    public function __construct(array $config)
    {
        if (!isset($config['port'])) {
            throw new Exception\InvalidArgumentException("Error missing required 'port' in config");
        }
        if (!(bool) filter_var($config['port'], FILTER_VALIDATE_INT) || $config['port'] < 1) {
            throw new Exception\InvalidArgumentException("Option 'port' must be numeric greater than 0");
        }
        $port = $config['port'];
        // Substitute magic vars is deprecated and will be removed in v1.0.0
        $config = array_merge(
            $this->getDefaultConfig($port),
            $this->substitutePlaceholders($config, $port)
        );
        $this->checkConfig($config);
        $this->config = $config;
    }

    /**
     * Return port on which standalone server listens.
     */
    public function getPort(): int
    {
        return (int) $this->config['port'];
    }

    /**
     * Return jar file of the server.
     */
    public function getServerJar(): string
    {
        return $this->config['server_jar'];
    }

    /**
     * Return java binary.
     */
    public function getJavaBin(): string
    {
        return $this->config['java_bin'];
    }

    /**
     * Return log file.
     */
    public function getLogFile(): string
    {
        return $this->config['log_file'];
    }

    /**
     * Return an array containing the java classpath(s) for the server.
     *
     * @return string[]
     */
    public function getClasspaths(): array
    {
        return $this->config['classpaths'];
    }

    public function getPidFile(): string
    {
        return $this->config['pid_file'];
    }

    /**
     * Return standalone server threads.
     *
     * @return int|string
     */
    public function getThreads()
    {
        return $this->config['threads'];
    }

    /**
     * Return standalone configuration.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return $this->config;
    }

    /**
     * Return standalone configuration.
     *
     * @deprecated use toArray() instead
     *
     * @return array<string, mixed>
     */
    public function getConfig(): array
    {
        return $this->toArray();
    }

    /**
     * Return default configuration options.
     *
     * @param int|string $port
     *
     * @return array<string, mixed>
     */
    protected function getDefaultConfig($port): array
    {
        return $this->substitutePlaceholders($this->default_config, $port);
    }

    /**
     * Substitute the placeholder {tcp_port} and {base_dir}
     * from a config array.
     *
     * @param array<string, mixed> $configArray associative array
     * @param int|string           $port
     *
     * @return array<string, mixed>
     */
    protected function substitutePlaceholders(array $configArray, $port): array
    {
        $substituted = [];
        $base_dir = $this->getBaseDir();

        foreach ($configArray as $key => $value) {
            if (is_string($value)) {
                $tmp = str_replace('{base_dir}', $base_dir, $value);
                $tmp = str_replace('{tcp_port}', (string) $port, $tmp);
            } else {
                $tmp = $value;
            }
            $substituted[$key] = $tmp;
        }

        return $substituted;
    }

    /**
     * Return pjbserver-tools installation base directory.
     *
     * @throws Exception\RuntimeException
     */
    public function getBaseDir(): string
    {
        // Four levels back.
        $ds = DIRECTORY_SEPARATOR;
        $dir = __DIR__ . "$ds..$ds..$ds..$ds..$ds";
        $base_dir = realpath($dir);
        if (!$base_dir) {
            $message = 'Cannot resolve project base directory.';
            throw new Exception\RuntimeException($message);
        }

        return $base_dir;
    }

    /**
     * Check configuration parameters.
     *
     * @throws Exception\InvalidArgumentException
     *
     * @param array<string, mixed> $config
     */
    protected function checkConfig(array $config): void
    {
        // Step 1: all required options
        $required = ['port', 'server_jar', 'log_file', 'pid_file', 'threads'];
        foreach ($required as $option) {
            if (!isset($config[$option]) || $config[$option] == '') {
                throw new Exception\InvalidArgumentException("Missing resuired configuration option: '$option''");
            }
        }

        // Step 2: server_jar file must exists
        if (!is_file($config['server_jar']) || !is_readable($config['server_jar'])) {
            throw new Exception\InvalidArgumentException("Server jar file not exists or unreadable. server-jar: '" . $config['server_jar'] . "'");
        }

        // Step 3: log and pid file should be creatable
        $temp_required_files = ['log_file', 'pid_file'];
        foreach ($temp_required_files as $option) {
            $file = $config[$option];
            $info = pathinfo($file);
            $dirname = $info['dirname'];
            if (!is_dir($dirname) || $dirname === '.') {
                $msg = "Option '$option' refer to an invalid or non-existent directory ($file)";
                throw new Exception\InvalidArgumentException($msg);
            }
            if (is_dir($file)) {
                $msg = "Option '$option' does not refer to a file but an existing directory ($file)";
                throw new Exception\InvalidArgumentException($msg);
            }
            if (file_exists($file) && !is_writable($file)) {
                $msg = "File specified in '$option' is not writable ($file)";
                throw new Exception\InvalidArgumentException($msg);
            }
        }

        // Step 4: Threads must be numeric greater than 0

        $threads = $config['threads'];

        if (!preg_match('/^([0-9])+$/', (string) $threads) || $threads <= 0) {
            $msg = "Parameter 'threads' must be valid integer greater than 0";
            throw new Exception\InvalidArgumentException($msg);
        }

        // Step 5: Java must be callable

        // @todo, many options exists

        // Step 6: Check classpaths autoload
        if (isset($config['classpaths'])) {
            if (!is_array($config['classpaths'])) {
                $msg = "Option 'classpaths' mus be a php array.";
                throw new Exception\InvalidArgumentException($msg);
            }
            foreach ($config['classpaths'] as $classpath) {
                if (preg_match('/\*\.jar$/', $classpath)) {
                    // Check if directory exists
                    $directory = preg_replace('/\*\.jar$/', '', $classpath);
                    if (!is_dir($directory) || !is_readable($directory)) {
                        $msg = "Classpath error, the directory of '$classpath' does not exists or is not readable";
                        throw new Exception\InvalidArgumentException($msg);
                    }
                } elseif (preg_match('/\.jar$/', $classpath)) {
                    // Check if file exists
                    if (!is_file($classpath) || !is_readable($classpath)) {
                        $msg = "Classpath error, the file '$classpath' does not exists or is not readable";
                        throw new Exception\InvalidArgumentException($msg);
                    }
                } else {
                    $msg = "Error in classpath, files to import must end by .jar extension ($classpath)";
                    throw new Exception\InvalidArgumentException($msg);
                }
            }
        }
    }
}
